netty/src/main/java/io/grpc/transport/netty/Utils.java

```diff
@@ -169,6 +170,18 @@ private static Http2Headers convertMetadata(Metadata headers) {
     return http2Headers;
   }
 
+  /**
+   * Should the channel be asked to flush this write. Checks if the calling thread is the event
+   * loop and assumes that a flush is following if that is the case. This is true for all event
+   * loop work following a read as onReadComplete triggers a flush. For other work scheduled
+   * onto the event loop outside of a read the caller (or event) must schedule a flush.
+   */
+  static boolean shouldFlush(Channel channel, boolean isLastWriteInSequence) {
+    // Don't flush is we are in the event loop as we have been triggered by a read event and
+    // a flush will occur on channel read completion.
+    return !channel.eventLoop().inEventLoop() && isLastWriteInSequence;
```

Comment:
As I mentioned before, this doesn't seem strict enough. Multiple channels will use the same event loop, so just being on the right event loop doesn't seem enough to know that channelReadComplete() will be called for _this_ channel.